File: go/scanner/module/ssltls/cve_2014_0160.go
package ssltls

import (
	"fmt"
	"os/exec"

	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/common/phaser/findings"
	"github.com/bloom42/phaser/scanner/module"
	"github.com/bloom42/phaser/scanner/module/ssltls/sslyze"
)

type CVE_2014_0160 struct{}

func (CVE_2014_0160) Name() string {
	return "ssltls/cve_2014_0160"
}

func (CVE_2014_0160) Description() string {
	return "Check for CVE-2014-0160 (a.k.a. heartbleed). See  http://heartbleed.com"
}

func (CVE_2014_0160) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (CVE_2014_0160) Version() string {
	return "0.1.0"
}

// TODO: better sslyze target handling
func (CVE_2014_0160) Run(scan *phaser.Scan, target *phaser.Target, port phaser.Port) (module.Result, []error) {
	errs := []error{}
	var ret module.Result

	if port.HTTP || !port.HTTPS {
		return ret, errs
	}

	host := fmt.Sprintf("%s:%d", target.Host, port.ID)
	command := "sslyze"
	commandArgs := []string{"--heartbleed", "--json_out=-", host}

	out, err := exec.Command(command, commandArgs...).Output()
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}

	sslyzeResult, err := sslyze.Parse(out)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}

	if len(sslyzeResult.AcceptedTargets) < 1 {
		return ret, errs
	}

	if sslyzeResult.AcceptedTargets[0].CommandsResults.Heartbleed.IsVulnerableToHeartbleed {
		ret = findings.URL{URL: "https://" + host}
	}

	return ret, errs
}

File: src/scanner/modules/ssltls/cve_2014_0160.rs
use crate::scanner::{
    module,
    findings,
    Scan,
    Target,
};
use std::process::{Command};
use crate::scanner::modules::ssltls::sslyze;


pub struct Cve2014_0160{}

impl module::BaseModule for Cve2014_0160 {
    fn name(&self) -> String {
        return String::from("ssltls/cve-2014-0160");
    }

    fn description(&self) -> String {
        return String::from("Check for CVE-2014-0160 (a.k.a. heartbleed). See http://heartbleed.com for more information");
    }

    fn author(&self) -> String {
        return String::from("Sylvain Kerkour <sylvain@kerkour.com>")
    }

    fn version(&self) -> String {
        return String::from("0.1.0");
    }
}

impl module::PortModule for Cve2014_0160 {
    fn run(&self, _: &Scan, target: &Target, port: &findings::Port) -> (Option<findings::Data>, Vec<String>) {
        let mut errs = vec!();
        let mut output = String::new();
        let mut ret = None;

        if !port.https {
            return (ret, errs);
        }

        let url = format!("{}:{}", &target.host, port.id);
        match Command::new("sslyze")
            .arg("--heartbleed")
            .arg("--json_out=-")
            .arg(&url)
            .output()
            {
            Ok(sslyze_output) => output = String::from_utf8_lossy(&sslyze_output.stdout).to_string(),
            Err(err)  => errs.push(format!("error executing sslyze: {}", err)),
        };

        if !output.trim().is_empty() {
            match serde_json::from_str::<sslyze::Scan>(&output) {
                Ok(sslyze_scan) => {
                    if sslyze_scan.accepted_targets.len() != 1 {
                        errs.push(
                            format!("wrong number of sslyze accepted_targets: expected 1, got: {}", sslyze_scan.accepted_targets.len())
                        );
                        return (ret, errs);
                    }
                    if sslyze_scan.accepted_targets[0].commands_results.heartbleed.is_vulnerable_to_heartbleed {
                        ret = Some(findings::Data::Url(findings::Url{
                            url: format!("https://{}", url),
                        }));
                    }
                },
                Err(err) =>  errs.push(format!("error parsing sslyze result: {}", err)),
            }
        }

        return (ret, errs);
    }
}

#[cfg(test)]
mod tests {
    use crate::scanner::module::BaseModule;

    #[test]
    fn module_name() {
        let module = super::Cve2014_0160{};
        assert_eq!("ssltls/cve-2014-0160", module.name());
    }
}
